fastlane env

hjanuschka opened the pull request:
- [x] atleast on unit test
- [x] documentation on how to use this feature, mostly in the ISSUE_TEMPLATE.md
- [x] Output should be nice markdown output, easy to copy and paste to GitHub
- [x] Instructions in the terminal on how to copy & paste to GitHub
- [x] Listing all the used Ruby gem versions (something similar to the Gemfile.lock, but has to be   fetched from the run time). It's easy to fetch the version of all gems that are being loaded
- [x] Xcode path (xcode-select -p)
- [x] Version of the Xcode installation on xcode-select -p
- [x] Print out the stack similar to CocoaPods (see below)
- [x] Executable path (similar to CocoaPods)
- [x] Loaded plugins (if any)

i took over https://github.com/fastlane/fastlane/pull/6252 - hope it is ok @KrauseFx 

Sample Output:
### Fastlane Environment
### Stack

| Key | Value |
| --- | --- |
| Fastlane | 1.105.2 |
| OS | 10.12 |
| Ruby | 2.2.1 |
| Xcode | /Applications/Xcode.app/Contents/Developer |
| Xcode Version | Xcode 8.0 Build version 8A218a |
| Git | git version 2.10.0 |
| Installation Source | /Users/hjanuschka/.rvm/gems/ruby-2.2.1/bin/fastlane |
| Host | Mac OS X 10.12 (16A323) |
| Ruby Lib Dir | /Users/hjanuschka/.rvm/rubies/ruby-2.2.1/lib |
### Fastlane Files:

`Fastfile` located in ./fastlane/Fastfile

```
lane :demo do
    puts "my fastfile"
end
```

`Appfile` located in ./fastlane/Appfile

```
my_app_file=1
```
### Fastlane Gems

| Gem | Version | Update-Status |
| --- | --- | --- |
| credentials_manager | 0.16.2 | 😇 Up-To-Date |
| fastlane_core | 0.52.1 | 😇 Up-To-Date |
| spaceship | 0.36.0 | 😇 Up-To-Date |
| cert | 1.4.3 | 😱 Update availaible |
| deliver | 1.14.2 | 😇 Up-To-Date |
| frameit | 2.8.0 | 😇 Up-To-Date |
| gym | 1.11.2 | 😇 Up-To-Date |
| sigh | 1.11.2 | 😇 Up-To-Date |
| match | 0.8.1 | 😇 Up-To-Date |
| pem | 1.3.2 | 😇 Up-To-Date |
| pilot | 1.10.1 | 😇 Up-To-Date |
| produce | 1.2.1 | 😇 Up-To-Date |
| scan | 0.13.1 | 😇 Up-To-Date |
| screengrab | 0.5.5 | 😇 Up-To-Date |
| snapshot | 1.16.2 | 😇 Up-To-Date |
| supply | 0.7.1 | 😇 Up-To-Date |
| fastlane | 1.105.2 | 😇 Up-To-Date |
### Loaded Fastlane Plugins:

| Plugin | Version | Update-Status |
| --- | --- | --- |
| fastlane-plugin-sharethemeal | 0.1.9 | 😇 Up-To-Date |
### Loaded Gems

| Gem | Version |
| --- | --- |
| rake | 11.3.0 |
| i18n | 0.7.0 |
| json | 1.8.3 |
| minitest | 5.9.1 |
| thread_safe | 0.3.5 |
| tzinfo | 1.2.2 |
| activesupport | 4.2.7.1 |
| addressable | 2.4.0 |
| ast | 2.3.0 |
| babosa | 1.0.2 |
| builder | 3.2.2 |
| bundler | 1.13.1 |
| byebug | 9.0.6 |
| colored | 1.2 |
| highline | 1.7.8 |
| commander | 4.4.0 |
| security | 0.1.3 |
| excon | 0.53.0 |
| gh_inspector | 1.0.2 |
| multi_json | 1.12.1 |
| plist | 3.1.0 |
| rubyzip | 1.1.7 |
| terminal-table | 1.4.5 |
| multipart-post | 2.0.0 |
| faraday | 0.9.2 |
| unf_ext | 0.0.7.2 |
| unf | 0.1.4 |
| domain_name | 0.5.20160826 |
| http-cookie | 1.0.3 |
| faraday-cookie_jar | 0.0.6 |
| faraday_middleware | 0.10.0 |
| fastimage | 1.6.8 |
| multi_xml | 0.5.5 |
| claide | 1.0.1 |
| cork | 0.2.0 |
| nap | 1.1.0 |
| open4 | 1.3.4 |
| claide-plugins | 0.9.2 |
| coderay | 1.1.1 |
| docile | 1.1.5 |
| simplecov-html | 0.10.0 |
| simplecov | 0.12.0 |
| tins | 1.12.0 |
| term-ansicolor | 1.4.0 |
| thor | 0.19.1 |
| coveralls | 0.8.15 |
| safe_yaml | 1.0.4 |
| crack | 0.4.3 |
| faraday-http-cache | 1.3.1 |
| git | 1.3.0 |
| sawyer | 0.7.0 |
| octokit | 4.3.0 |
| redcarpet | 3.3.4 |
| danger | 0.10.1 |
| diff-lcs | 1.2.5 |
| dotenv | 2.1.1 |
| fakefs | 0.8.1 |
| mini_magick | 4.5.1 |
| rouge | 1.11.1 |
| xcpretty | 0.2.4 |
| net-ssh | 3.2.0 |
| net-sftp | 2.1.2 |
| krausefx-shenzhen | 0.14.10 |
| slack-notifier | 1.5.1 |
| xcpretty-travis-formatter | 0.0.4 |
| jwt | 1.5.6 |
| little-plugger | 1.1.4 |
| logging | 2.1.0 |
| memoist | 0.15.0 |
| os | 0.9.6 |
| signet | 0.7.3 |
| googleauth | 0.5.1 |
| httpclient | 2.8.2.4 |
| hurley | 0.2 |
| mime-types | 1.25.1 |
| uber | 0.0.15 |
| representable | 2.3.0 |
| retriable | 2.1.0 |
| google-api-client | 0.9.15 |
| terminal-notifier | 1.7.1 |
| word_wrap | 1.0.0 |
| xcode-install | 2.0.7 |
| xcodeproj | 1.3.2 |
| rest-client | 1.6.9 |
| stm_api | 0.1.6 |
| fastlane-plugin-sharethemeal | 0.1.9 |
| method_source | 0.8.2 |
| parser | 2.3.1.4 |
| powerpack | 0.1.1 |
| slop | 3.6.0 |
| pry | 0.10.4 |
| pry-byebug | 3.4.0 |
| rainbow | 2.1.0 |
| rspec-support | 3.1.2 |
| rspec-core | 3.1.7 |
| rspec-expectations | 3.1.2 |
| rspec-mocks | 3.1.3 |
| rspec | 3.1.0 |
| rspec_junit_formatter | 0.2.3 |
| ruby-progressbar | 1.8.1 |
| unicode-display_width | 1.1.1 |
| rubocop | 0.38.0 |
| webmock | 1.19.0 |
| yard | 0.8.7.6 |

KrauseFx committed:
[WIP] Hemal add environment printer for GitHub issues

Commit:
enhance env print out

KrauseFx commented on fastlane/lib/fastlane/environment_printer.rb:
```diff
@@ -0,0 +1,207 @@
+      Gem.loaded_specs.values.each do |x|
+        update_status = "N/A"
+
+        next unless Fastlane::Actions::UpdateFastlaneAction.all_tools.include?(x.name)
```
That's not a good dependency to have. Instead you can just access `Fastlane::TOOLS`

KrauseFx commented on fastlane/lib/fastlane/actions/update_fastlane.rb:
```diff
@@ -25,6 +25,9 @@ class UpdateFastlaneAction < Action
         "match"
       ]
+      def self.all_tools
+        ALL_TOOLS
```
Please remove that, see other comment

KrauseFx commented on fastlane/lib/fastlane/commands_generator.rb:
```diff
@@ -14,9 +14,12 @@ def self.start
+      # do not use "include" as it may be some where in the commandline where "env" is required, therefore explicit index->0
+      unless ARGV[0] == "env"
+        # *after* loading the plugins
+        Fastlane.plugin_manager.load_plugins
```
Nice idea, what's the reason we don't run it?

hjanuschka replied:
actually this one prints the loaded plugins table - and we cannot set a env on `:env` command,
as `self.start` is run before commander! - if we don't suppress this, then the `Terminal-Table`  
of the loaded plugins will be included when using `fastlane env|pbcopy` - wich would mess up issues.
the plugins later on will be loaded in the `environment_printer` - with a `ENV` to suppress table :smile:

KrauseFx commented on fastlane/lib/fastlane/commands_generator.rb:
```diff
@@ -240,6 +243,16 @@ def run
 
+      command :env do |c|
+        c.syntax = 'fastlane env'
+        c.description = 'Print your fastlane environment'
```
> Print your fastlane environment, use this when you submit an issue on GitHub

KrauseFx commented on fastlane/lib/fastlane/environment_printer.rb:
```diff
@@ -0,0 +1,207 @@
+          update_url = FastlaneCore::UpdateChecker.generate_fetch_url(plugin)
+          latest_version = FastlaneCore::UpdateChecker.fetch_latest(update_url)
+          if Gem::Version.new(installed_version) == Gem::Version.new(latest_version)
+            update_status = "😇 Up-To-Date"
```
Maybe we should use ✅  and 🚫  instead 👍

KrauseFx commented on fastlane/lib/fastlane/environment_printer.rb:
```diff
@@ -0,0 +1,207 @@
+
+    def self.print_loaded_fastlane_gems
+      # Fastlanes own gems
+      env_output = "### Fastlane Gems\n\n"
```
fastlane should always be lower case, also on the comment above

KrauseFx replied:
Gems is lower case here too

KrauseFx commented on fastlane/lib/fastlane/environment_printer.rb:
```diff
@@ -0,0 +1,207 @@
+      env_output
+    end
+  end
+  class MarkdownTableFormatter
```
You need a space after an `end`

KrauseFx replied:
The class shouldn't be here, please move this in a separate file instead

KrauseFx commented on fastlane/spec/env_spec.rb:
```diff
@@ -0,0 +1,16 @@
+        to_return(status: 200, body: '{"version": "0.16.2",  "status": "ok"}', headers: {})
+    end
+    it "Prints Env Dump" do
+      expect(Fastlane::EnvironmentPrinter.get).to include("Fastlane")
```
Could we do some more assertions here?

KrauseFx commented on fastlane/spec/env_spec.rb:
```diff
@@ -0,0 +1,16 @@
+        with(headers: { 'Host' => 'fastlane-refresher.herokuapp.com:443', 'User-Agent' => 'excon/0.53.0' }).
+        to_return(status: 200, body: '{"version": "0.16.2",  "status": "ok"}', headers: {})
+    end
+    it "Prints Env Dump" do
```
Please add a space before this

KrauseFx commented:
Awesome work on this, thanks. I added some comments

hjanuschka commented:
ok thx for review:

  * changed emojis
  * fastlane lowercase everwhere, also the one gem typo
  * switched to: `Fastlane::TOOLS` awesome, didn't know that.
    * removed modifications in `update_fastlane.rb`
  * MarkdownTableFormatter in own file
  * wording in `commands_generator`
  * space before `it` in spec
  * added a few more assertions in `spec` - also added `generated at` to the markdown, and in spec check if unit-tests reaches end :smile:
  

 looks now like:


```ruby
fastlane env
```

### fastlane Environment

### Stack

| Key                 | Value                                        |
| ------------------- | -------------------------------------------- |
| fastlane            | 1.105.2                                      |
| OS                  | 10.12                                        |
| Ruby                | 2.2.5                                        |
| Xcode               | /Applications/Xcode.app/Contents/Developer   |
| Xcode Version       | Xcode 8.0 Build version 8A218a               |
| Git                 | git version 2.8.4 (Apple Git-73)             |
| Installation Source | /Users/hja/.rvm/gems/ruby-2.2.5/bin/fastlane |
| Host                | Mac OS X 10.12 (16A323)                      |
| Ruby Lib Dir        | /Users/hja/.rvm/rubies/ruby-2.2.5/lib        |


### fastlane Files:

`Fastfile` located in ./fastlane/Fastfile

```
lane :demo do
	puts "asdf"
end
```


`Appfile` located in ./fastlane/Appfile

```
app_file=1
```


### fastlane gems

| Gem        | Version | Update-Status       |
| ---------- | ------- | ------------------- |
| spaceship  | 0.36.0  | ✅ Up-To-Date        |
| cert       | 1.4.3   | ✅ Up-To-Date        |
| deliver    | 1.14.2  | ✅ Up-To-Date        |
| frameit    | 2.8.0   | ✅ Up-To-Date        |
| gym        | 1.11.2  | 🚫 Update availaible |
| sigh       | 1.11.2  | ✅ Up-To-Date        |
| match      | 0.8.1   | ✅ Up-To-Date        |
| pem        | 1.3.2   | ✅ Up-To-Date        |
| pilot      | 1.10.1  | ✅ Up-To-Date        |
| produce    | 1.2.1   | ✅ Up-To-Date        |
| scan       | 0.13.1  | ✅ Up-To-Date        |
| screengrab | 0.5.5   | ✅ Up-To-Date        |
| snapshot   | 1.16.2  | ✅ Up-To-Date        |
| supply     | 0.7.1   | ✅ Up-To-Date        |
| fastlane   | 1.105.2 | ✅ Up-To-Date        |


### Loaded fastlane Plugins:

| Plugin                       | Version | Update-Status       |
| ---------------------------- | ------- | ------------------- |
| fastlane-plugin-sharethemeal | 0.1.0   | 🚫 Update availaible |


### Loaded Gems

| Gem                          | Version      |
| ---------------------------- | ------------ |
| rake                         | 11.3.0       |
| i18n                         | 0.7.0        |
| json                         | 1.8.3        |
| minitest                     | 5.9.1        |
| thread_safe                  | 0.3.5        |
| tzinfo                       | 1.2.2        |
| activesupport                | 4.2.7.1      |
| addressable                  | 2.4.0        |
| ast                          | 2.3.0        |
| babosa                       | 1.0.2        |
| builder                      | 3.2.2        |
| bundler                      | 1.13.4       |
| byebug                       | 9.0.6        |
| colored                      | 1.2          |
| highline                     | 1.7.8        |
| commander                    | 4.4.0        |
| security                     | 0.1.3        |
| credentials_manager          | 0.16.2       |
| excon                        | 0.53.0       |
| gh_inspector                 | 1.0.2        |
| multi_json                   | 1.12.1       |
| plist                        | 3.1.0        |
| rubyzip                      | 1.1.7        |
| terminal-table               | 1.4.5        |
| fastlane_core                | 0.52.1       |
| multipart-post               | 2.0.0        |
| faraday                      | 0.9.2        |
| unf_ext                      | 0.0.7.2      |
| unf                          | 0.1.4        |
| domain_name                  | 0.5.20160826 |
| http-cookie                  | 1.0.3        |
| faraday-cookie_jar           | 0.0.6        |
| faraday_middleware           | 0.10.0       |
| fastimage                    | 1.6.8        |
| multi_xml                    | 0.5.5        |
| spaceship                    | 0.36.0       |
| cert                         | 1.4.3        |
| claide                       | 1.0.1        |
| cork                         | 0.2.0        |
| nap                          | 1.1.0        |
| open4                        | 1.3.4        |
| claide-plugins               | 0.9.2        |
| coderay                      | 1.1.1        |
| docile                       | 1.1.5        |
| simplecov-html               | 0.10.0       |
| simplecov                    | 0.12.0       |
| tins                         | 1.12.0       |
| term-ansicolor               | 1.4.0        |
| thor                         | 0.19.1       |
| coveralls                    | 0.8.15       |
| safe_yaml                    | 1.0.4        |
| crack                        | 0.4.3        |
| faraday-http-cache           | 1.3.1        |
| git                          | 1.3.0        |
| sawyer                       | 0.7.0        |
| octokit                      | 4.3.0        |
| redcarpet                    | 3.3.4        |
| danger                       | 0.10.1       |
| deliver                      | 1.14.2       |
| diff-lcs                     | 1.2.5        |
| dotenv                       | 2.1.1        |
| fakefs                       | 0.8.1        |
| mini_magick                  | 4.5.1        |
| frameit                      | 2.8.0        |
| rouge                        | 1.11.1       |
| xcpretty                     | 0.2.4        |
| gym                          | 1.11.2       |
| net-ssh                      | 3.2.0        |
| net-sftp                     | 2.1.2        |
| krausefx-shenzhen            | 0.14.10      |
| sigh                         | 1.11.2       |
| match                        | 0.8.1        |
| pem                          | 1.3.2        |
| pilot                        | 1.10.1       |
| produce                      | 1.2.1        |
| slack-notifier               | 1.5.1        |
| xcpretty-travis-formatter    | 0.0.4        |
| scan                         | 0.13.1       |
| screengrab                   | 0.5.5        |
| snapshot                     | 1.16.2       |
| jwt                          | 1.5.6        |
| little-plugger               | 1.1.4        |
| logging                      | 2.1.0        |
| memoist                      | 0.15.0       |
| os                           | 0.9.6        |
| signet                       | 0.7.3        |
| googleauth                   | 0.5.1        |
| httpclient                   | 2.8.2.4      |
| hurley                       | 0.2          |
| mime-types                   | 1.25.1       |
| uber                         | 0.0.15       |
| representable                | 2.3.0        |
| retriable                    | 2.1.0        |
| google-api-client            | 0.9.15       |
| supply                       | 0.7.1        |
| terminal-notifier            | 1.7.1        |
| word_wrap                    | 1.0.0        |
| xcode-install                | 2.0.7        |
| xcodeproj                    | 1.3.2        |
| fastlane                     | 1.105.2      |
| fastlane-plugin-sharethemeal | 0.1.0        |
| method_source                | 0.8.2        |
| parser                       | 2.3.1.4      |
| powerpack                    | 0.1.1        |
| slop                         | 3.6.0        |
| pry                          | 0.10.4       |
| pry-byebug                   | 3.4.0        |
| rainbow                      | 2.1.0        |
| rest-client                  | 1.6.9        |
| rspec-support                | 3.1.2        |
| rspec-core                   | 3.1.7        |
| rspec-expectations           | 3.1.2        |
| rspec-mocks                  | 3.1.3        |
| rspec                        | 3.1.0        |
| rspec_junit_formatter        | 0.2.3        |
| ruby-progressbar             | 1.8.1        |
| unicode-display_width        | 1.1.1        |
| rubocop                      | 0.38.0       |
| webmock                      | 1.19.0       |
| yard                         | 0.8.7.6      |



 *generated at:* **17/10/2016 22:19**

hjanuschka commented:
btw: once this is merged - https://github.com/fastlane/fastlane/pull/6586 i would love to add this to the env_printer too - whatever is merged first, maybe as a following PR 👍

Commit:
fix crash if no internet-connection is availaible - as on :airplane:

KrauseFx commented:
Upgraded the design a bit 👍 

<img width="821" alt="screenshot 2016-10-20 11 17 06" src="https://cloud.githubusercontent.com/assets/869950/19572349/d25c6ae0-96b6-11e6-98fc-31d627440223.png">
<img width="781" alt="screenshot 2016-10-20 11 17 02" src="https://cloud.githubusercontent.com/assets/869950/19572348/d25be778-96b6-11e6-813e-d0807f5422b9.png">
<img width="777" alt="screenshot 2016-10-20 11 15 45" src="https://cloud.githubusercontent.com/assets/869950/19572347/d25a0c28-96b6-11e6-9aed-daa2c60a0013.png">
<img width="797" alt="screenshot 2016-10-20 11 15 37" src="https://cloud.githubusercontent.com/assets/869950/19572350/d25e353c-96b6-11e6-8b8d-eb0c6cdaed71.png">

KrauseFx commented:
Moved to https://github.com/fastlane/fastlane/pull/6649
